### DISCUSSION TITLE
midi throgh arduino?

### DISCUSSION THREAD

**Discussion by souravlal** [marcel-licence/esp32_midi_sampler#18] [Q&A] [answered]
Hy, I don't have a midi controller. can I send general midi signals using an arduino nano to control esp32 audio kit, with arduino midi library?

**Comment by marcel-licence** [comment]
Yes you can use another controller outputting serial 3.3V signals.

**Reply by jazzbiddy** [reply]
Hi @marcel-licence 
Can you explain more? What pins are you using on the esp32 audio kit to connect to the other arduino? Is the RX pin IO18 on the top right header? I'm trying to understand this statement in the config.h:
``` 
#ifdef ESP32_AUDIO_KIT
#define MIDI_RX_PIN 18
//#define MIDI_SERIAL2_BAUDRATE   115200
#else
#define MIDI_RX_PIN 16
//#define TXD2 17
#endif
```

and how that relates to this Midi_Setup() in midi_interface.ino
```
void Midi_Setup()
{
#ifdef MIDI_RX_PIN
#ifdef TXD2
    Serial2.begin(MIDI_SERIAL2_BAUDRATE, SERIAL_8N1, MIDI_RX_PIN, TXD2);
#else
    Serial2.begin(MIDI_SERIAL2_BAUDRATE, SERIAL_8N1, MIDI_RX_PIN);
#endif
    pinMode(MIDI_RX_PIN, INPUT_PULLUP); /* can be connected to open collector output */
#endif
}
```

There are 2 scenario's I'm trying to run this in. 
1) When the esp32 audio kit is connected to PC using usb(uart) to test midi messages coming in from hairless midi to trigger samples
2) When the esp32 audio kit is connected to a teensy using the teensy's hardware serial2 ports (RX9/TX10) to trigger samples 
 https://www.pjrc.com/teensy/td_libs_MIDI.html

Sorry if this was answered in another post. Any help on the wiring to the esp32 audio kit is appreciated. And thanks for sharing your project and creating this discussion area.
-jazzyb

**Reply by marcel-licence** [reply]
@jazzbiddy So you can leave "midi_interface.ino" as it is. The configuration is in config.h. Using the ESP32 Audio Kit would require to have the line `#define ESP32_AUDIO_KIT` in this file. Then MIDI_RX_PIN will be mapped to IO18 on the pin header of the ESP32 Audio Kit. You need to connect this pin (Rx) to the Tx pin of your other hardware (ensure level is 3.3V) and GND should be connected (may be also connected via USB indirectly).

With "MIDI_RECV_FROM_SERIAL" defined you can send MIDI messages over the UART which appears on your computer when the ESP32 Audio Kit is connected via USB.

The ESP32 does not send any MIDI data so there is no TX line used.

**Reply by jazzbiddy** [reply]
@marcel-licence I'll give a try and let you know how it works out. Thanks for the fast response.

**Reply by souravlal** [reply]
@marcel-licence
I have gone through the 'z_config.ino', in that code you have mapped the midi controller input messages.
But if I use an Arduino Nano with 'midi library' it will send different values isn't it?
You have used hexadecimal values in your code.
How can I map my Arduino nano to control few buttons and few pots? Whichever note or control doesn't matter just for testing out.
Can you please help?

**Reply by marcel-licence** [reply]
For example 
`    { 0x0, 0x11, "S1", NULL, Sampler_SetADSR_Attack, 0},`
is linked to a control change message on channel (0x0) =0, data1 = 0x11 = 17.

So the message containing three bytes would be 0xB0 0x11 <VAL> (which can be a value between 0..127)

Sending the lowest value can be done with 0xB0 0x11 0x00 and the highest 0xB0 0x11 0x7F.
I am using speedcrunch to convert hex to decimal.

There are also some control change messages linked two buttons. In that case you can send the 0x7F for a button down and 0x00 for button up.

Maybe I will find some time to make a tutorial. Sorry for the confusion

**Reply by souravlal** [reply]
I want to connect two buttons which should act as normal keyboard buttons(note), and two pots which should control any two control change( like reverb, pitch bend, Attack, etc...).
Which should be connected to an Arduino nano which should send MIDI messages to ESP32 audio kit.
What messages should I send?

**Reply by jazzbiddy** [reply]
@souravlal I think you need to choose the messages that you want to send since your arduino nano is basically a custom midi controller. Once you decide what midi messages you want to send, then you just make sure that you modify Marcel's code to correspond your midi messages, instead of the ones he used with is controller.

@marcel-licence you have some great youtube videos showing your code while in use - I think what would be helpful is a walkthrough of the different modules (tabs) in your project calling out the parts that users can change for their own setup. Of course, this depends on how much you expect others to know before we try to use your code. 

BTW - I did get sounds coming out with your help, but it doesn't sound like the "pet_bottle.wav" file.  Also when I set this line to `1`
```
    /* use this to easily test the output */
#if 1
    Sampler_NoteOn(0, 64, 1);
#endif
```
It also doesn't sound like the file.  I get a sound, so I'm happy that I'm making progress, but I do feel like I'm going to have to remove a lot of what you have and start with a basic example of simple sounds playing from midi notes.  then add back in all of your advanced stuff.

**Reply by souravlal** [reply]
Is there no way to make it universal?
At the moment if I want to use a MIDI keyboard I have to program it according to the keyboard? If I want to use another keyboard I will have to change the code again right?
But whatever commercial ones I have seen it will work with all keyboards.
Also There is a 'sampler box' based on Raspberry pi in which we install the OS for sampler box in an SD card and put it in the raspberry pi and connect a USB drive with samples, which also states it will work with all  MIDI keyboards.
Also all PC software synthesizers are just plug and play.
As far as I know MIDI messages have a universal standard, right?
Do anyone have an idea?

@marcel-licence Is it possible to add a keyboard mapping function which will ask to map the buttons,pots,modulation wheel,etc.. to the control we want in the ESP32 audio kit?
You are already using a display right?  Won't it be nice to add the code which will ask to set control changes to map at booting up of the ESP32 audio kit in the display? 
Or may be use another micro-controller to do the function (receive the MIDI signal and ask to map it to ESP32 audio kit using display). Otherwise it would be a pain to modify the code each time you want to use a different MIDI controller. Will it effect the latency? 
This is my idea I am presenting in front of you because I know you are a very good programmer.

**Reply by marcel-licence** [reply]
In general it would be possible to add MIDI learn but it would require a display then. I don't know which controllers you are using. Mine comes with a little tool to setup the pots, sliders, buttons etc:
![grafik](https://user-images.githubusercontent.com/65245426/126523878-a7520dad-db31-4b3a-8b15-62fe2479e943.png)
For each element the control change is active and you can setup a control number (data1/2nd byte), the channel (second nibble of the first byte) and the range (data2/3rd byte).

**Comment by Slider2732** [comment]
A very interesting discussion. Thanks all for pointers to what I thought was just my own lack of knowledge of CC.
It would seem that I'm in a similar boat... struggling to change the code to match MIDI controller CC outputs. Am also using an Arduino to convert USB MIDI to 5 pin. 

An example USB output from the Worlde Easycontrol 9 is like this when moving slider 1 upward, using 'amidi -p hw:3,0,0 -d' on Linux Mint:
B0 03 24
B0 03 89
B0 03 95 etc
When moving slider 2 it may be:
B0 04 15
B0 04 23
B0 04 35 etc
When moving slider 3 it may be:
B0 05 23
B0 05 51
B0 05 78 etc

So, simple enough, 0xB0 is CC, the next number would be data1 and the third would be data2. At least, that's what I thought. 
Using an Arduino with a USB shield on top, I get great results connecting the Worlde's USB only output to it and the 5 pin DIN MIDI cable from the Arduino shield to my Roland Juno-06A. Filter, slider levels, pot movements can be controlled. Volume is volume, slider and pot levels change as expected. 
Software is from here: https://github.com/YuuichiAkagawa/USBH_MIDI
The factory set CC's are being output from the unmodified Worlde (there's no editing software for Linux that I know of, only Win and Mac).
It therefore caused some confusion to only hear random changes from the sampler code. I would think I'd changed the code correctly, but would hear only random note firings or a sort of pitch shift effect. There must be 40 different uploads of those changes on the poor audio kit lol

Here's the standard: https://professionalcomposers.com/midi-cc-list/ 
As such, it's likely the controllers are indeed plug and play to several parameters, 'out of the box'. 

The changes Marcel described above make total sense, but I don't get total sense output when implemented. However, the new code with screen support hasn't been tried yet... there may be unseen changes. Will check it out  :)

**Reply by marcel-licence** [reply]
In this case you could setup for example:

`struct midiControllerMapping myMapping[] =
{
    /* slider */
    { 0x0, 0x03, "Slider1", NULL, Sampler_SetADSR_Attack, 0},
    { 0x0, 0x04, "Slider2", NULL, Sampler_SetADSR_Decay, 1},
    { 0x0, 0x05, "Slider3", NULL, Sampler_SetADSR_Sustain, 2},
}

struct midiMapping_s midiMapping =
{
    NULL,
    Sampler_NoteOn,
    Sampler_NoteOff,
    Sampler_PitchBend,
    Sampler_ModulationWheel,
    myMapping, // refer to your MIDI map here
    sizeof(myMapping) / sizeof(myMapping[0]), // provide size of your MIDI map here
};`

In that case you should be able to control attack, decay, sustain

**Reply by marcel-licence** [reply]
This may help https://github.com/marcel-licence/esp32_midi_sampler/blob/main/midi_setup.md

**Comment by Slider2732** [comment]
Came back to post the link, having seen your reply elsewhere in other discussions. 
:)
THANKS for the extra info for the Worlde controller!

Hopefully your info page also helps the other 2 posters. 


Update -
Works superbly. My big error was not having 0x0 leading the midi Mapping lines.
Have added Release and Reverb. Only full Release produces the glitchy stuff with chords, an original problem with my own build. 
Screen code works too - I hope Adafruit change their ST7735 library. Will incorporate your code submit to fix the offset on the 160x80 (exact same display). I think my resistor numbers have to be edited in where you set up the button presses as only Key1 and sometimes Key2 work at the moment. Issues are just tiny changes. Am delighted!

**Reply by marcel-licence** [reply]
Oh sounds good. I am very happy to hear that the screen works. Adafruit was quick with fixing their Neopixel library (a pull request from me went into their lib). But it seems that there is no progress with the display. Also other people provided possible fixes (pull requests).
I've also commit a change where I could remove some glitch in the sound. If you like you can also provide the problematic sample if you like -> you can open an issue also for this.

**Comment by Slider2732** [comment]
Am not hearing any glitches so far, now that the Release can be set with the MIDI controller. 
However, I can't seem to load a new sample, am stuck with the initial one at switch on. It may be my Key range and likely is. All other keys seem to work fine. 
EDIT - fixed
The screen doesn't have the bottom row displayed, which your screen init likely fixes. Haven't fixed it yet due to messing about with reverb and delay... plus playing house piano lines (at last !!!!) haha 
EDIT - your Adafruit submit works great. No offset anymore.   
Delightfully, the pitch range control is enough to move played notes up or down by something like an octave or so in either direction. Very handy for my 3 octave keyboard set up. 

All in all, for others reading and wondering, a very cheap Worlde controller and Arduino+USB shield works very well. It may answer the OP's question.


Btw, you need to create a 'Buy Me A Coffee' Paypal button somewhere!

**Reply by souravlal** [reply]
@marcel-licence, So you are saying that playing a key (note) and controlling pitch bend and modulation wheel is common and we can use any MIDI keyboard with the ESP32 audio kit to control these parameters without any modification in the code. Right? 
But for more options you have added in the project like ADSR control and Reverb, etc will require the control message to be mapped accordingly. Right?
If that is the case I can send a MIDI message to play a note with an Arduino using MIDI library in Arduino IDE. Right?

Won't it be rather better to connect control knobs directly to the ESP32 audio kit itself to control ADSR, Reverb and such parameters, without the help of control knobs and buttons of the MIDI keyboard only use it to play notes and pitch bend? Which will make this project more universal and can be used with any MIDI keyboard.
Will this be possible?  If yes can you please let me know what should be the tweaks in the code?

**Reply by souravlal** [reply]
Please somebody help me out with my above question. 👆

**Reply by marcel-licence** [reply]
Or a "Buy me a display" button xD. Unfortunately I got the wrong one. I would try one of the nextion display but they sell wrong displays using the wrong title :-/ You will also find youtube videos refering to them https://youtu.be/R0iB4coRr_o <- this is not correct

**Comment by Slider2732** [comment]
Your idea of it is sort of correct.
Mainly, the MIDI library will send the right note on and note off messages without issue. 
It will work 100% to fire samples at the correct pitch. There are a couple of problems though, one of which is to make sure you don't use channel 1 to fire notes. The code is set up to put single hit samples across a MIDI keyboard on channel 1 and most Arduino MIDI guides will have channel 1 as the MIDI output channel. All you'll get is a single same pitched note, repeated across the keyboard!
Just have to set the output channel as 2.
The other problem is of polyphony being truncated if the standard Release time is used. If Release is not at full you can get 7 or 8 notes and zero issues. Getting Release to change however has required my MIDI controller to work. Every note played on its own works, but pressing chord keys fast can end up with missing notes. The fix is to send a Release CC message of say 80 (out of 127). 

CC messages are a little more involved. Marcel's new page on CC implementation in the repo explains it.
What I have done, is to keep his Edirol controller namings and mapping names, but change the values at the start of each line. For example, in z_config: He may have 
   { 0x11, 0x51, "rec", NULL, Sampler_Record, 0},
My own MIDI controller outputs 0xB0 0x17 on the button I wanted to use for the function, so the line becomes
   { 0x0, 0x17, "rec", NULL, Sampler_Record, 0},

Using an Arduino to output CC similarly can be found with a search or three. NotesAndVolts has a good guide:
https://www.youtube.com/watch?v=DXhxdsGREsU
https://www.youtube.com/watch?v=AYx55nj5vSo

A main reason for not having the functions on pots is the need for pins... the Audio Kit is sparse. 
But, there's nothing saying you couldn't write some code for an Arduino that used pots and buttons, which then sent the data over MIDI to the Audio Kit. Was half way done doing that myself, until I got the MIDI controller going.

**Reply by souravlal** [reply]
@Slider2732  You said that need to change the release parameter less than maximum. Can't this be done inside the code itself by changing the initial value of release.
Can't the missing key problem when pressing chord keys faster be rectified somehow?
You said that using an Arduino to send the cc messages? Can this be done parallel, I mean connect a MIDI keyboard and an Arduino to the ESP32 audio kit at the same time in parallel? Will it cause any issues? Aren't the MIDI CC messages send only when changing knobs or pressing buttons, or the messages are send continuously?

**Reply by marcel-licence** [reply]
You can do this after a sample has been loaded, also at startup and that can be all coded.
The Arduino + MIDI keyboard would not work at the same time with the Audio Kit. You could for example connect the keyboard to your arduino and implement some MIDI merge function or use some library which can combine internal messages with incoming.
CC messages are usually only sent when changing knobs, pots etc. So you will only see updated values without repetitions.

**Reply by souravlal** [reply]
Since the cc messages are not sent repeatedly, and when controlling cc messages using an Arduino we won't do anything with the MIDI keyboard(won't sent any messages) Arduino+MIDI keyboard wouldn't work?
MIDI merge function with a micro-controller would be a good option but won't it add to the latency?

**Reply by souravlal** [reply]
@Slider2732 'marcel's new page about cc' where is it I couldn't find it. Can you give me the link?

**Comment by Slider2732** [comment, marked as answer]
Some good questions, I like the exploration because it's learning I should have done 30 years back!
Marcel's willingness to explore such questions is another reason why I think us guys should contribute to his Starbucks bills  :)
I agree with the code changes, but could never achieve them on the Release code sections. It was noticed when the controller started working. 

There can be troubles when using 2 MIDI devices at exactly the same time, such as holding down a looped sample and moving a slider. The held note from the Arduino, the slider on the controller. I've had the latest code lock up, panic and reset doing that (only had the latest code work with my MIDI controller). Also ran into issues when adding an extra MIDI In to my Korg X5DR, by way of 2 MIDI connectors with wires linked between the pins. 

Generally, several devices 'could' be on the same input line, if they don't send messages together. 
On a synth build here, I have an Arduino spitting out MIDI NoteOn's and NoteOff's from a keyboard bed - notes go to Marcel's Simple Synth, the Hypersaw variation of it and an unrelated Eminent/ARP Solina string synth converted to MIDI input on an Arduino. All connect to the same MIDI line from the Arduino firing the keybed's notes.
The 'trick' is to make each synth circuit the end of the MIDI line... lockups happen otherwise and would need merge code.  
Also, as I understand it, the fan out ability of Arduino TX is something like TTL rather than CMOS. Strong enough to trigger several devices. Star configurations are often cited as a way to cut latency, no Thru stuff  :)

Marcel may be pleased to learn that the bizarre PAiA 3 octave keybed has been changed to a street find freebie of a Yamaha Clavinova CLP123 for the sampler project. 
The current system here, is an Arduino Nano sending MIDI NoteOn/Off from the (broken electronics) Clavinova to the Audio Kit. The Worlde EasyControl 9 plugs into an Arduino MIDI shield, with Arduino Uno. Output is to the Audio Kit as well. Works fine. 
Best practice, is to move sliders or press buttons on the controller, then play some notes. It's natural enough.
